src/extra/components/scan.py

```diff
@@ -0,0 +1,293 @@
+
+    ```python
+            -----------------------------------------------------------
+    In [1]: |s = Scan(run["MOTOR/MCMOTORYFACE"])                      |
+            |s                                                        |
+            -----------------------------------------------------------
+    Out[1]: Scan over MOTOR/MCMOTORYFACE.actualPosition from -0.5800 to -0.5600:
+             Steps: 21
+             Scan time: 0:17:54
+             Average step length: 511.33 trains (51.13s)
+             Average step size: 1.00e-03
+
+            Detection parameters:
+             resolution: 5.00e-04
+             min_trains: 235
+
+            ----------------------------------------------------------
+    In [2]: |s.plot() # Check the detection results                  |
+            ----------------------------------------------------------
+    ```
+    """
+
+    def __init__(self, motor, resolution=None, min_trains=None):
+        """
+        The class tries to detect the best parameters automatically, but it will
+        still fail in some situations. If that happens either the `resolution`
+        or `min_trains` parameters (unlikely to be both) will need to be set
+        manually, typically on a per-motor rather than per-scan basis. You
+        should always check the results of the detection with
+        [Scan.plot()][extra.components.Scan.plot] to verify that they're
+        correct.
+
+        The `resolution` parameter is the trickiest one to guess
+        automatically. For example, if it's set too high then multiple real
+        steps will be detected as single steps. And if it's set too low and
+        there's a lot of drift in the motor steps, there can be multiple
+        detected steps within a single real step.
+
+        `min_trains` rarely fails badly. Usually you will only need to tweak
+        this if the length of the steps varies a lot and some of the smaller
+        steps are being cut out (this often happens with energy scans).
+
+        Args:
+            motor (SourceData, KeyData, DataArray): An
+                object that contains the motor positions and train IDs. If it's a
+                `SourceData` object the `actualPosition` key will be used. Note that
+                if a [xarray.DataArray][] is passed it *must* have a `trainId`
+                coordinate with the train IDs (i.e. obtained with
+                [KeyData.xarray()][extra_data.KeyData.xarray]). Raw Numpy arrays are not
+                allowed.
+            resolution (float): The usable resolution of the motor. Ideally this
+                should be slightly above the noise level of the motor (and below the
+                step size of the scan). It will be guessed if not passed explicitly.
+            min_trains (int): The minimum number of trains per-step in the scan. It
+                will be guessed if not passed explicitly.
+        """
+        self._name = "motor"
```

Comment:
When an `xarray.DataArray` is passed, it may be nice to set this argument manually via an optional argument or try to take it from the `DataArray`'s name (`KeyData.xarray()` will set this to `{source}.{key}` by default).

Reply:
Ah nice, didn't know about `DataArray.name`. Added in 7f69671.